pkg/yurthub/healthchecker/health_checker.go

```diff
@@ -21,114 +21,368 @@ import (
-		healthyThreshold:  healthyThreshold,
+func (hcm *healthCheckerManager) newLease(base *coordinationv1.Lease, checker *checker) *coordinationv1.Lease {
+	var lease *coordinationv1.Lease
+	if base == nil {
+		lease = &coordinationv1.Lease{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      hcm.holderIdentity,
+				Namespace: corev1.NamespaceNodeLease,
+			},
+			Spec: coordinationv1.LeaseSpec{
+				HolderIdentity:       pointer.StringPtr(hcm.holderIdentity),
+				LeaseDurationSeconds: pointer.Int32Ptr(hcm.leaseDurationSeconds),
+			},
+		}
+	} else {
+		lease = base.DeepCopy()
 	}
 
-	initHealthyStatus, err := PingClusterHealthz(c.healthzClient, c.serverHealthzAddr)
-	if err != nil {
-		klog.Errorf("cluster(%s) init status: unhealthy, %v", c.serverHealthzAddr, err)
+	lease.Spec.RenewTime = &metav1.MicroTime{Time: hcm.clock.Now()}
+
+	if lease.OwnerReferences == nil || len(lease.OwnerReferences) == 0 {
+		if node, err := checker.client.CoreV1().Nodes().Get(hcm.holderIdentity, metav1.GetOptions{}); err == nil {
+			lease.OwnerReferences = []metav1.OwnerReference{
+				{
+					APIVersion: corev1.SchemeGroupVersion.WithKind("Node").Version,
+					Kind:       corev1.SchemeGroupVersion.WithKind("Node").Kind,
+					Name:       hcm.holderIdentity,
+					UID:        node.UID,
+				},
+			}
+		} else {
+			klog.Errorf("failed to get node %q when trying to set owner ref to the node lease: %v", hcm.holderIdentity, err)
+		}
+	}
+
+	return lease
+}
+
+func (hcm *healthCheckerManager) storageLease(lease *coordinationv1.Lease) error {
+	if lease == nil {
+		return nil
 	}
-	c.clusterHealthy = initHealthyStatus
-	if c.clusterHealthy {
-		metrics.Metrics.ObserveServerHealthy(c.netAddress, 1)
+
+	accessor := meta.NewAccessor()
+	accessor.SetKind(lease, coordinationv1.SchemeGroupVersion.WithKind("Lease").Kind)
+	accessor.SetAPIVersion(lease, coordinationv1.SchemeGroupVersion.String())
+
+	cacheLeaseKey := fmt.Sprintf(cacheLeaseKeyFormat, hcm.holderIdentity)
+	oldObj, err := hcm.sw.Get(cacheLeaseKey)
+	if err == nil && oldObj != nil {
+
+		oldRv, err := accessor.ResourceVersion(oldObj)
+		if err != nil {
+			klog.Errorf("failed to get old object resource version for %s, %v", cacheLeaseKey, err)
+			return err
+		}
+
+		newRv, err := accessor.ResourceVersion(lease)
+		if err != nil {
+			klog.Errorf("failed to get new object resource version for %s, %v", cacheLeaseKey, err)
+			return err
+		}
+
+		oldRvInt, _ := strconv.Atoi(oldRv)
+		newRvInt, _ := strconv.Atoi(newRv)
+		if newRvInt <= oldRvInt { // resource version is incremented or not
+			return nil
+		}
+
+		return hcm.sw.Update(cacheLeaseKey, lease)
+	} else if os.IsNotExist(err) || oldObj == nil {
+		return hcm.sw.Create(cacheLeaseKey, lease)
 	} else {
-		metrics.Metrics.ObserveServerHealthy(c.netAddress, 0)
+		if err != storage.ErrStorageAccessConflict {
+			return hcm.sw.Create(cacheLeaseKey, lease)
+		}
+		return err
+	}
+}
+
+func (hcm *healthCheckerManager) getChecker() *checker {
+	hcm.RLock()
```

Comment:
how about delete lock of `healthCheckerManager`?

Reply:
fixed